### ISSUE TITLE
[question] How are multiple remotes resolved?

### ISSUE THREAD

**Issue by Daniel-Roberts-Bose** [conan-io/conan#6036]
So I have a question on how conan handles multiple remotes. 

I have two remotes defined in my configuration, lets call them `R1` and `R2`. I have some package called `PKG1` that exists in both `R1` and `R2`, however, the version of `PKG1` in `R1` is more up to date 

`R1` -> `PKG1/v2.0.0@User/Channel`
`R2` -> `PKG1/v1.0.0@User/Channel`

I run a `conan install -r R1` and its finding the package at version 2, but then for some reason its searching `R2` for that package and not finding it, causing the command to error out and fail with a 404 error

Does conan search for the package in both remotes or just the one specified? Its possible that this is a bug where its checking the wrong remote, even though I'm telling the install command to check `R1`

TL:DR
I am telling conan to look in one remote for a package, but its looking in the other remote and erroring out when it doesn't find it at the given version

This is is being run under conan version 1.20.1

As another side note, the issue may be unrelated to each remote containing a different version, however, why is it even checking `R2` when I have explicitly told it to search in `R1`. The exact error being displayed is `ERROR: 404: Not Found. [Remote: R2]`

**Comment by jgsogo** [comment]
Hi @Daniel-Roberts-Bose 

I'm trying to reproduce your issue, but I think I'm going to need some more help from your side.

This is my current setup:

```bash
⇒  conan --version
Conan version 1.20.1
[python3]jgsogo@MacBook-Pro-de-Javier:~/dev/conan/issues/6036|
⇒  conan search PKG1\*
There are no packages matching the 'PKG1*' pattern
[python3]jgsogo@MacBook-Pro-de-Javier:~/dev/conan/issues/6036|
⇒  conan remote list
conan-center: https://conan.bintray.com [Verify SSL: True, Disabled: True]
bincrafters: https://api.bintray.com/conan/bincrafters/public-conan [Verify SSL: True, Disabled: True]
jgsogo: https://api.bintray.com/conan/jgsogo/conan-packages [Verify SSL: True, Disabled: True]
conan-server1: http://0.0.0.0:9301/ [Verify SSL: True]
conan-server2: http://0.0.0.0:9302/ [Verify SSL: True]
```

So, I have to remotes enabled, and there is no local package matching the `PKG1` pattern

In the remotes:

```bash
⇒  conan search PKG1\* -r conan-server1
Existing package recipes:

PKG1/v2.0.0
[python3]jgsogo@MacBook-Pro-de-Javier:~/dev/conan/issues/6036|
⇒  conan search PKG1\* -r conan-server2
Existing package recipes:

PKG1/v1.0.0
```

And if I run the `install` it works:

```bash
⇒  conan install PKG1/v2.0.0@ -r conan-server1         
Configuration:
[settings]
arch=x86_64
...

PKG1/v2.0.0: Retrieving from server 'conan-server1' 
PKG1/v2.0.0: Trying with 'conan-server1'...
Downloading conanmanifest.txt: 100%|##########| 58.0/58.0 [00:00<00:00, 116kB/s]
Downloading conanfile.py: 100%|##########| 115/115 [00:00<00:00, 319kB/s]
PKG1/v2.0.0: Downloaded recipe revision 0
Installing package: PKG1/v2.0.0
Requirements
    PKG1/v2.0.0 from 'conan-server1' - Downloaded
Packages
    PKG1/v2.0.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download

PKG1/v2.0.0: Retrieving package 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 from remote 'conan-server1' 
Downloading conanmanifest.txt: 100%|##########| 59.0/59.0 [00:00<00:00, 117kB/s]
Downloading conaninfo.txt: 100%|##########| 150/150 [00:00<00:00, 436kB/s]
Downloading conan_package.tgz: 100%|##########| 63.0/63.0 [00:00<00:00, 181kB/s]
Decompressing conan_package.tgz: 100%|##########| 63.0/63.0 [00:00<00:00, 42.7kB/s]
PKG1/v2.0.0: Package installed 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
PKG1/v2.0.0: Downloaded package revision 0
```

Is your output different? Do you have already `PKG1/v2.0.0` in your cache associated to `R2`?

**Comment by Daniel-Roberts-Bose** [comment]
Hi @jgsogo 

Yeah looking at the output I have, it could be some sort of caching issue. The package is coming from cache and was not downloaded. 

What is the expected behavior if a recipe was previously downloaded from one remote (`conan-server2`), and now you are telling it to use another remote (`conan-server1`)?

**Comment by jgsogo** [comment]
The general rule: if you force one remote in the command line, Conan should use just that remote. There is a risk if you are not using revisions and you are downloading binaries from different remotes, as the recipes used and uploaded to those remotes might be different, and hence, the binary package could have been generated by different recipes. That's the reason we associate each recipe with a remote, to not mix binaries.

When using revisions, as the revision is the hash of the recipe (+ packaged sources), it is safe to download binaries from different remotes if all of them have been generated with the same recipe revision... because if the recipe and sources are the same, the binary should be "the same" (different configurations, different options,... but same sources).

**Comment by Daniel-Roberts-Bose** [comment]
Thanks for some clarification on this, it helps a lot

We haven't migrated to using revisions yet so this might explain some of the weirdness, however, if the recipe specifies the remote it came from, could that still cause some weirdness?

From my perspective, I would expect if a recipe was downloaded from one remote but you are trying to update from a different remote, it would notice the difference and download the recipe from the different remote. It sounds like you are saying that this is not actually the case, and it will simply use the recipe if the version is the same, is that correct?

**Comment by jgsogo** [comment]
Without revisions, it will.

If you already have a recipe in your cache, Conan will use it to compute the package_id and then it will go to the defined remote to look for the package. No extra check is done about the recipe. So you may end-up with:
 * `conan install name/version@issue/testing -r remote1`:
    * recipe from `remote1`
    * a package from `remote1` 
 * `conan install name/version@issue/testing -r remote2 -s <other_setting>`:
    * a different package from `remote2`

Yes, you will mix packages that might be generated using different recipe revisions, but **as revisions were disabled, both of them have `revision=0` and match**. You may inspect the file `.../data/name/version/issue/testing/metadata.json` in the cache to see the remote associated with each item.

With revisions enabled, Conan detect that the recipe revision is different, and removes the existing packages from the previous revision.

**Comment by AbrilRBS** [comment]
Closing as solved in Conan 2, where revisions are mandatory, and the `--update` flag can be used to control wether Conan will search thru all the remotes for newer versions
